shared/test_run_query.go

```diff
@@ -152,44 +156,76 @@ func (t testRunQueryImpl) LoadTestRunKeys(
 			if ids, err = loadKeysForRevision(t.store, query, product.Revision); err != nil {
 				return nil, err
 			}
-			revKeyFilter := mapset.NewSet()
+			revIDFilter := mapset.NewSet()
 			for _, id := range ids {
-				revKeyFilter.Add(id)
+				revIDFilter.Add(id)
 			}
-			productKeyFilter = merge(productKeyFilter, revKeyFilter)
+			log.Debugf("Found %v keys for %s@%s", revIDFilter.Cardinality(), product.BrowserName, product.Revision)
+			productIDFilter = merge(productIDFilter, revIDFilter)
 		}
 		if product.BrowserVersion != "" {
-			var versionKeys mapset.Set
-			if versionKeys, err = loadKeysForBrowserVersion(t.store, query, product.BrowserVersion); err != nil {
+			var versionIDs mapset.Set
+			if versionIDs, err = loadKeysForBrowserVersion(t.store, query, product.BrowserVersion); err != nil {
 				return nil, err
 			}
-			productKeyFilter = merge(productKeyFilter, versionKeys)
-		}
-		// TODO(lukebjerring): Indexes + filtering for OS + version.
-		query = query.Order("-TimeStart")
-
-		if from != nil {
-			query = query.Filter("TimeStart >=", *from)
-		}
-		if to != nil {
-			query = query.Filter("TimeStart <", *to)
+			log.Debugf("Found %v keys for %s", versionIDs.Cardinality(), product.BrowserVersion)
+			productIDFilter = merge(productIDFilter, versionIDs)
 		}
 
+		// If we have a specific set of possibilities, it's much cheaper to
+		// turn the query on its head (filter the entities).
```

Comment:
One more nit: this function is quite big now, especially the part below which has too many levels of nesting. Let's pull the two branches of the if switch out into two functions.

Reply:
Done (for the client-side filtering only).